doc/adr/0005-wego-core-auth-strategy.md

```diff
@@ -0,0 +1,125 @@
+- [Bitbucket Implicit Grant](https://developer.atlassian.com/cloud/bitbucket/oauth-2/#2--implicit-grant--4-2-)
+  - No `CLIENT_SECRET` required here
+
+### Kubernetes Cluster
+
+All operations will need to go through an API server that is installed on the cluster via the `wego gitops install` command. This API server will run with a service account that provides the ability to do the following:
+
+- Create `Namespace` resources
+- Create `Application`, `Source`, and `Kustomize|Helm` resources
+- Read Kubernetes objects in the `wego-system` namespace
+
+### Security Concerns
+
+Each of the Git Providers listed above support some sort of "personal access token" that allow for "impersonation" style authorization. These tokens never expire, and provide "sudo" access to the user's Github account.
+
+For this reason, we want to utilize the more short-lived OAuth tokens to avoid exposing a personal access token. We also don't want to persist these tokens longer than we have to.
+
+In addition, flows that require a `CLIENT_SECRET` will require us to embed it in the binary, which leaves us open to exposing it via decompilation.
+
+### UX Concerns
+
+Given that the Weave GitOps UI and API server may be running on a user's cluster, we may not always know the OAuth callback URL ahead of time. For this reason, we will need to use OAuth flows that do not require a static `redirect_uri`. The flows listed above in the Git Providers section fit that constraint.
+
+We will need to ensure that the `redirect_uri` is configurable by the user via environment variables. This will allow users to run and expose the UI/API servers on whichever endpoint they choose.
+
+## Design
+
+Since most modern Git Providers will support a form of browser-based OAuth, **we will utilize the browser for both the CLI and UI authentication with Git Providers**.
+
+In the case of the CLI, we can utilize a short-lived browser session that will recieve the OAuth callback and pass the token to the CLI.
+
+**Note that we can skip the OAuth portion of the flow if a user provides a `--token=<some value>` flag to the CLI command:**
+
+![CLI Auth Diagram](cli_auth.svg)
+
+UI auth will work in a similar way, with a more straight-forward set of steps:
+
+![UI Auth Diagram](ui_auth.svg)
+
+TLDR:
+
+1. The user does OAuth with their Git Provider
+2. We use the resulting access_token to push a Deploy Key to the repository (giving us `git` permissions)
+3. We create a pull request for the repository via the Git Provider HTTP API
+
+For browser security, we will convert the Git Provider OAuth token to a JSON Web Token (JWT) to protect against Cross-site Scripting (XSS) attacks. The encrypted JWT will allow a malicious script to authenticate with the Weave GitOps API only, whereas passing the unencrypted OAuth token to the browser would allow a malicious script to authenticate with the Github API.
```

Comment:
Would `wego` generate a keypair for encrypting the JWT, if so, how and where is that stored?

Reply:
Good question. Can we generate that at runtime? Every time the app restarts, the user would have to re-authenticate.

Outside of generating at runtime, it would need to work like all of the other individually-identifiable information: injected into the binary by CI.

Reply:
Added a section on this.